[CodeGen] Port AtomicExpand to new Pass Manager

Ris-Bali opened the pull request:
Port the `atomicexpand` pass to the new Pass Manager. 
Fixes #64559

boomanaiden154 commented:
@Ris-Bali Are you still working on this?

Ris-Bali commented:
Yes I am!!

arsenm commented on llvm/include/llvm/CodeGen/AtomicExpand.h:
```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+/// \file
+///
+///
+///
+//===----------------------------------------------------------------------===//
```
Drop empty comment block

arsenm commented on llvm/include/llvm/CodeGen/AtomicExpand.h:
```diff
@@ -0,0 +1,35 @@
+
+} // end namespace llvm
+
+#endif // LLVM_CODEGEN_ATOMICEXPAND_H
```
Missing newline at end of file?

MaskRay commented:
What's the rationale behind the pass renaming?

Ris-Bali commented on llvm/lib/Passes/PassRegistry.def:
```diff
@@ -429,6 +429,7 @@ FUNCTION_PASS("strip-gc-relocates", StripGCRelocates())
 FUNCTION_PASS("structurizecfg", StructurizeCFGPass())
 FUNCTION_PASS("tailcallelim", TailCallElimPass())
 FUNCTION_PASS("typepromotion", TypePromotionPass(TM))
+FUNCTION_PASS("expand-atomic", ExpandAtomicPass(TM))
```
Will rebase the PR and make this change.

arsenm commented on llvm/lib/Passes/PassRegistry.def:
```diff
@@ -429,6 +429,7 @@ FUNCTION_PASS("strip-gc-relocates", StripGCRelocates())
 FUNCTION_PASS("structurizecfg", StructurizeCFGPass())
 FUNCTION_PASS("tailcallelim", TailCallElimPass())
 FUNCTION_PASS("typepromotion", TypePromotionPass(TM))
+FUNCTION_PASS("atomicexpand", AtomicExpandPass(TM))
```
Renaming the pass shouldn't be done in the same change as the new PM support

Ris-Bali replied:
Reverted back to original name.

arsenm commented on llvm/lib/CodeGen/AtomicExpandPass.cpp:
```diff
@@ -340,7 +342,39 @@ bool AtomicExpand::runOnFunction(Function &F) {
+  if (skipFunction(F))
+    return false;
+
+  auto *TPC = getAnalysisIfAvailable<TargetPassConfig>();
+  if (!TPC)
+    return false;
```
I don't remember why we can't just require this or make the TargetMachine a pass constructor item like in new PM

arsenm commented on llvm/lib/CodeGen/AtomicExpandPass.cpp:
```diff
@@ -340,7 +342,39 @@ bool AtomicExpand::runOnFunction(Function &F) {
+    return PreservedAnalyses::all();
+
+  PreservedAnalyses PA;
+  PA.preserveSet<CFGAnalyses>();
```
This isn't true if a cmpxchg expansion is performed

arsenm commented on llvm/lib/CodeGen/AtomicExpandPass.cpp:
```diff
@@ -340,7 +342,33 @@ bool AtomicExpand::runOnFunction(Function &F) {
 
-bool AtomicExpand::bracketInstWithFences(Instruction *I, AtomicOrdering Order) {
+bool AtomicExpandLegacy::runOnFunction(Function &F) {
+  if (skipFunction(F))
+    return false;
```
This should not use skipFunction, it's a mandatory lowering pass. It's not part of the current code

arsenm commented on llvm/lib/CodeGen/AtomicExpandPass.cpp:
```diff
@@ -340,7 +342,31 @@ bool AtomicExpand::runOnFunction(Function &F) {
 
-bool AtomicExpand::bracketInstWithFences(Instruction *I, AtomicOrdering Order) {
+bool AtomicExpandLegacy::runOnFunction(Function &F) {
+  i
```
Stray character

arsenm commented on llvm/lib/Target/PowerPC/PPCTargetMachine.cpp:
```diff
@@ -457,7 +457,7 @@ TargetPassConfig *PPCTargetMachine::createPassConfig(PassManagerBase &PM) {
   if (TM->getOptLevel() != CodeGenOptLevel::None)
     addPass(createPPCBoolRetToIntPass());
-  addPass(createAtomicExpandPass());
+  addPass(createAtomicExpandLegacyPass());
```
We should probably just add this to the default pipeline (not in this change)

Ris-Bali replied:
Will create a patch for this one this gets merged

arsenm commented on llvm/lib/Target/NVPTX/NVPTXTargetMachine.cpp:
```diff
@@ -380,7 +380,7 @@ void NVPTXPassConfig::addIRPasses() {
     addStraightLineScalarOptimizationPasses();
   }
 
-  addPass(createAtomicExpandPass());
```
we probably should just put this in the standard pass pipeline instead of forcing every target to add this

Ris-Bali replied:
What I understand is I need to add this to the TargetPassconfig::addIRPasses() function which is called by every target. Could you please confirm if this is ryt ? There doesn't seem to be a default pipeline for codegen passes in the new PM. Please do correct me if I am wrong.

arsenm replied:
it's a separate change from this, not related

arsenm commented on llvm/lib/CodeGen/AtomicExpandPass.cpp:
```diff
@@ -340,7 +342,31 @@ bool AtomicExpand::runOnFunction(Function &F) {
+  AtomicExpandImpl AE;
+
+  bool Changed = AE.run(F, TM);
+  if (!Changed)
```
could figure out if CFG changes were made, but that's a more complex change which should be done separately

Ris-Bali commented:
All the tests pass now.

Ris-Bali commented:
Should I revert the formatting changes in `llvm/include/llvm/CodeGen/Passes.h` most of the formatting changes done are not related to this PR but were still being detected by `check_code_formatter` so I had to comply.

aeubanks commented:
in the future it would be nice to split out the renaming into a separate PR

Ris-Bali commented:
> > Should I revert the formatting changes in `llvm/include/llvm/CodeGen/Passes.h` ? most of the formatting changes done are not related to this PR but were still being detected by `check_code_formatter` so I had to comply.
> 
> Sorry for the unclear request, just use `git clang-format` to format the original patch.

The problem is the test only passes if Passes.h is fully formatted (the tests fail with the latest commit where I reverted the formatting changes" originally I have changed a single line in that file. git clang-format is currently formatting the entire file. It may be a bug although I am not sure

Ris-Bali committed:
[NewPM] Port AtomicExpandPass to new PassManager

Ris-Bali commented:
> > > > Should I revert the formatting changes in `llvm/include/llvm/CodeGen/Passes.h` ? most of the formatting changes done are not related to this PR but were still being detected by `check_code_formatter` so I had to comply.
> > > 
> > > 
> > > Sorry for the unclear request, just use `git clang-format` to format the original patch.
> > 
> > 
> > The problem is the formatting tests only passes if Passes.h is fully formatted([prev_commit](https://github.com/llvm/llvm-project/pull/71220/commits/02b535096e47c1f7f00acd31dee507adf268a89c)) (the tests fail with the latest commit where I reverted the formatting changes" originally I have changed a single line in that file. git clang-format is currently formatting the entire file. It may be a bug although I am not sure
> 
> You can squash the change here.

Have squashed the changes.
